### DISCUSSION TITLE
How to check "no such window: target window already closed" in VBA code?

### DISCUSSION THREAD

**Discussion by GHRyunosuke** [GCuser99/SeleniumVBA#96] [Q&A] [answered]
In case I manually close the browser that is started by SeleniumVBA (by clicking the right upper corner "X"). And then I continue to use the "driver" to navigate to some URL (i.e. driver.NavigateTo XXX ) it will fail and i got the message in immediate window. 

![image](https://github.com/GCuser99/SeleniumVBA/assets/25561705/f8f5029b-bfa4-425b-89fe-5b872f68f413)

So, is there a way to check if the browser is closed? or more specifically to check if "no such window: target window already closed" programmatically in VBA?

By the way, for now, this is how i handle this situation. Wonder if there is a more elegant way to do this.

```
    On Error Resume Next
    driver.NavigateTo strURL
    If Err.Number <> 0 Then
        Set driver = Nothing
        Set driver = New WebDriver
        driver.StartChrome
        driver.OpenBrowser
        driver.NavigateTo strURL
    End If
   On Error GoTo 0`
```

**Comment by 6DiegoDiego9** [comment]
Hi @GHRyunosuke, I would suggest you create a specific procedure like NavigateToFailsafe containing the error management logic, checking for the specific Err.Number and/or Err.Description.

**Comment by GCuser99** [comment]
@GHRyunosuke, in addition to @6DiegoDiego9's great suggestion, in your error block I don't think you need to shutdown and then start back up the driver object - try this instead:
```vba
    On Error Resume Next
    driver.NavigateTo strURL
    If Err.Number <> 0 Then
        driver.CloseBrowser
        driver.OpenBrowser
        driver.NavigateTo strURL
    End If
    On Error GoTo 0
```
And I don't know if it is any better than the approach that you are currently taking, but here is an alternative way using two windows (see [Managing Windows](https://github.com/GCuser99/SeleniumVBA/wiki#managing-windows) in the Wiki for more info):
```vba
    driver.StartChrome
    driver.OpenBrowser
    
    'create and activate another window - call it the "navigation" window
    driver.Windows.SwitchToNew svbaWindow
    
    driver.NavigateTo "https://www.google.com/"

    If driver.Windows.Count = 1 Then
        'the navigation window was closed without using window.CloseIt method
        'Selenium WebDriver thinks the closed window is active so we need
        'to activate the open spare window
        driver.Windows.SwitchToNext
        'then (optionally) open and activate a new "navigation" window 
        '(if needed - could just use the existing window too!)
        driver.Windows.SwitchToNew svbaWindow
    End If

    driver.NavigateTo "https://www.google.com/"

    driver.CloseBrowser
    driver.Shutdown
```
PS: I do not know of a way to detect the situation where a window gets closed non-programmatically like in your example, other than trapping the resulting error. For the browser session to remain valid, there must be at least one open window, as far as I know.

**Comment by GHRyunosuke** [comment, marked as answer]
Thanks a lot, @6DiegoDiego9 and @GCuser99. I think I will take the easier approach combining both your advice as follows to handle this situation.

```
Sub NavigateToFailsafe(driver As WebDriver, strURL As String)

    On Error Resume Next
    driver.NavigateTo strURL
    
     If Err.Number <> 0 Then
        driver.CloseBrowser
        driver.OpenBrowser
        driver.NavigateTo strURL
     End If
     
    On Error GoTo 0
    
End Sub
```
